### PR TITLE
=all #13783 replace protobuf dependency by embedded version

### PR DESCRIPTION
**Pull request by patriknw**

### CODE REVIEW

**Comment by patriknw** [comment]
The first commit is a rebase of https://github.com/akka/akka/pull/16896

The second commit makes ProtobufSerializer independent of protobuf version
* well, as long as they provide the parseFrom and toByteArray
* It is using reflection to find the `parseFrom` and `toByteArray` methods to avoid
  dependency to `com.google.protobuf`.
* also special case com.google.protobuf when loading serialization binding
* migration guide

**Comment by patriknw** [comment]
Refs #13783

File: project/AkkaBuild.scala [hunk 1]
```diff
@@ -87,7 +87,7 @@ object AkkaBuild extends Build {
   lazy val actorTests = Project(
     id = "akka-actor-tests",
     base = file("akka-actor-tests"),
-    dependencies = Seq(testkit % "compile;test->test")
+    dependencies = Seq(testkit % "compile;test->test", protobuf)
```
Comment (patriknw) [inline project/AkkaBuild.scala]: It doesn't look like this is needed

**Comment by hepin1989** [comment]
for it's a embead version ,why not just based on 2.6.1?

**Comment by patriknw** [comment]
Refs  #13783

**Comment by ktoso** [comment]
> for it's a embead version ,why not just based on 2.6.1?

Would that be binary compatible with Akka 2.3 which used this version of protobuf?

**Comment by hepin1989** [comment]
I am using protobuf 2.6.1 and it evicted protobuf 2.5.0,I have not encounter any probleam,but not that sure for a larger scope.

**Comment by patriknw** [comment]
Moving to 2.6.1 would probably be great (assuming that it is wire compatible) if this was done a month ago. Might be risky to do it now, since we are already on bonus time for 2.4.0. Does it solve any issue or increase performance dramatically?

**Comment by patriknw** [comment]
ah, forgot mima

**Comment by hepin1989** [comment]
seems not ,and here is the (changelog)[https://github.com/google/protobuf/blob/master/CHANGES.txt].
anyway ,embed is a good start.

**Comment by patriknw** [comment]
failure was unrelated know issue #18176

**Comment by patriknw** [comment]
PLS BUILD

**Comment by patriknw** [comment]
I did two manual tests to make sure that I didn't messed up anything.

1. Persistent replay of messages stored with old version.

2. Cluster with old and new nodes sending plain protobuf messages (and thereby using the ProtobufSerializer) between them.

All good!

**Comment by drewhk** [comment]
LGTM

**Comment by patriknw** [comment]
rebased

**Comment by ktoso** [comment]
Changes look safe to me, LGTM

**Pull request merged**